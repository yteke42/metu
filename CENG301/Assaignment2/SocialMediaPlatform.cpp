#include "SocialMediaPlatform.h"
#include "User.h"
#include "SortedLinkedList.h"
#include "LinkedListNode.h"

using namespace std;

// Implement functions in the "SocialMediaPlatform.h"
SocialMediaPlatform::SocialMediaPlatform() : users(new SortedLinkedList()) {}

SocialMediaPlatform::~SocialMediaPlatform() { delete users; }

bool SocialMediaPlatform::addUser(const int userId, const std::string name) {
    if (users->checkIfExist(userId)) return false;

    User *newUser = new User();

    newUser->setId(userId);
    newUser->setName(name);
    users->add(newUser);
    return true;
} 

bool SocialMediaPlatform::removeUser(const int userId) {
    // Iterate through all users in the platform
    LinkedListNode *currentUserNode = users->getHead();
    while (currentUserNode != nullptr) {
        // Remove the user to be deleted from the current user's followeds and followers list
        currentUserNode->u->removeFollowed(userId);
        currentUserNode->u->removeFollower(userId);

        // Move to the next user
        currentUserNode = currentUserNode->next;
    }

    // Remove the user from the platform
    return users->remove(userId);
}





bool SocialMediaPlatform::followUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (firstUserNode->u->getFolloweds()->checkIfExist(secondUserId)) {
        // User is already following the second user
        return false;
    }

    firstUserNode->u->addFollowed(secondUserNode->u);
    secondUserNode->u->addFollower(firstUserNode->u);
    return true;
}


bool SocialMediaPlatform::unfollowUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (!firstUserNode->u->getFolloweds()->remove(secondUserId) ||
        !secondUserNode->u->getFollowers()->remove(firstUserId)) {
        return false; // Either user not found or user not followed, return false
    }

    return true;
}


void SocialMediaPlatform::showFollowersOf(const int userId) const {
    LinkedListNode *userNode = users->getPointerTo(userId);
    if (userNode != nullptr) {
        std::cout << "Users following the user with ID " << userId << ":" << std::endl;
        if (!userNode->u->getFollowers()->isEmpty()) {
            userNode->u->getFollowers()->print();
        } else {
            std::cout << "None" << std::endl;
        }
    } else {
        std::cout << "Cannot show. There is no user with ID " << userId << "." << std::endl;
    }
}





void SocialMediaPlatform::showAllUsers() const {
    std::cout << "Users in the social media platform:" << std::endl;
    if (users->isEmpty()) {
        std::cout << "None" << std::endl;
    } else {
        users->print();
    }
}





