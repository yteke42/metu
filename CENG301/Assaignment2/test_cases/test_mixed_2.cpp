#include<iostream>
#include<string>
#include "SocialMediaPlatform.h"

using namespace std;

int main() {

    string truth_table[2] = {"false", "true"};

    SocialMediaPlatform SMP;

    cout<<truth_table[SMP.addUser( 3, "Lonely Artist" )]<<endl;
    cout<<truth_table[SMP.addUser( 2, "Eater Man" )]<<endl;
    cout<<truth_table[SMP.addUser( 4, "Pet Lover" )]<<endl;   
    cout<<truth_table[SMP.addUser( 4, "Musician Guy" )]<<endl;
    cout<<truth_table[SMP.removeUser( 4 )]<<endl;
    cout<<truth_table[SMP.addUser( 4, "Cyborg" )]<<endl;
    cout<<truth_table[SMP.addUser( 1, "Gamer" )]<<endl;
    cout<<truth_table[SMP.addUser( 1, "Ali" )]<<endl;
    cout<<truth_table[SMP.removeUser( 7 )]<<endl;
    cout<<truth_table[SMP.addUser( 6, "Sportsman" )]<<endl;
    cout<<endl;

    SMP.showAllUsers();
    cout<<endl;

    cout<<truth_table[SMP.followUser( 1, 6 )]<<endl;
    cout<<truth_table[SMP.followUser( 1, 4 )]<<endl;
    cout<<truth_table[SMP.followUser( 2, 3 )]<<endl;
    cout<<truth_table[SMP.followUser( 2, 4 )]<<endl;
    cout<<truth_table[SMP.followUser( 3, 1 )]<<endl;
    cout<<truth_table[SMP.followUser( 1, 3 )]<<endl;
    cout<<truth_table[SMP.followUser( 1, 3 )]<<endl;
    cout<<truth_table[SMP.followUser( 1, 7 )]<<endl;
    cout<<endl;

    cout<<truth_table[SMP.unfollowUser( 2, 3 )]<<endl;
    cout<<truth_table[SMP.unfollowUser( 2, 6 )]<<endl;
    cout<<truth_table[SMP.unfollowUser( 1, 7 )]<<endl;
    cout<<endl;

    SMP.showFollowersOf(1);
    cout<<endl;
    SMP.showFollowersOf(2);
    cout<<endl;
    SMP.showFollowersOf(3);
    cout<<endl;
    SMP.showFollowersOf(4);
    cout<<endl;
    SMP.showFollowersOf(5);
    cout<<endl;
    SMP.showFollowersOf(6);
    cout<<endl;
    SMP.showFollowersOf(7);
    cout<<endl;

    return 0;
}
