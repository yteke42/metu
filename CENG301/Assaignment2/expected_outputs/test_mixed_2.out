true
true
true
false
true
true
true
false
false
true

Users in the social media platform:
1, Gamer
2, Eater Man
3, Lonely Artist
4, Cyborg
6, Sportsman

true
true
true
true
true
true
false
false

true
false
false

Users following the user with ID 1:
3, Lonely Artist

Users following the user with ID 2:
None

Users following the user with ID 3:
1, Gamer

Users following the user with ID 4:
1, Gamer
2, Eater Man

Cannot show. There is no user with ID 5.

Users following the user with ID 6:
1, Gamer

Cannot show. There is no user with ID 7.

